### PR TITLE
Endpoint for scheduling a job for generating translations using AI4Bharat model and Google Translate 

### PR DESCRIPTION
**Pull request by prakharrathi25**
# Description

Add a new endpoint for generating translations from a dataset instance of type `SentenceText` and create a new dataset instance of type `TranslationPair`. The translations are made using the ULCA IndicTrans models and the Google Translate Model. The translations are performed in batch format. 

The endpoint will accept the `input_dataset_instance_id` , and a list of `languages` for which it needs to create output instances along with the `output_dataset_instance_id`. 

Fixes [SN-250](https://project-sunbird.atlassian.net/browse/SN-250?atlOrigin=eyJpIjoiY2Y4NWVlZGVkNWNjNGMwZGFkNDE3MDVjYjIxNzI5ZTAiLCJwIjoiaiJ9) and [SN-245](https://project-sunbird.atlassian.net/browse/SN-245?atlOrigin=eyJpIjoiYWI1MDUwN2IwNjA5NGEzZGI3NjFiMjM2ZjAwNmQ4OTQiLCJwIjoiaiJ9) 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Tested the API endpoint on Swagger with the 'Try it out' feature
- [x] Any other tests you performed for checking edge cases.
- [x] Any other tests you performed for checking exceptions.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

## Checklist for API

If you added new api endpoints / made changes to exsting endpoints, please fill this checklist. 

- [x] The endpoint is accessible through Swagger.
- [x] The endpoint supports permissions and authentication for different roles.
- [x] All exceptions have been handled and appropriate status code is returned to the user.
- [x] I have added docstrings for ViewSets / Serializers.

## Additional Information 

The google translate endpoint requires GCP Translation API Credentials. Contact @GokulNC or @prakharrathi25 for the same. 

## Format 

The request-response format is shown below. 

Indic Trans endpoint
image

Google Translate Endpoint
image

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,7 @@
 all_data
 SentenceText-2022-05-01.csv
 wiki-full-data.csv
+ai4bharat-indic-nlp-f0df53305a6f.json
```
Comment (GokulNC) [inline .gitignore:4]: What if we change the keys in the future? :)

I guess `*.json` is OK.
Similarly `*.csv` above

File: backend/users/utils.py [hunk 2]
```diff
@@ -1,5 +1,156 @@
...
+    "Santali": "sat",
+    "Sindhi": "sd",
+    "Sinhala": "si",
+    "Tamil": "ta",
+    "Telugu": "te",
+    "Urdu": "ur",
+}
+
+# Language code to language name GOOGLE TRANSLATE
+LANG_CODE_TO_NAME_GOOGLE = {
```
Comment (GokulNC) [inline backend/users/utils.py:37]: You can just do the following to avoid duplicate code:

```py
LANG_CODE_TO_NAME_GOOGLE = {
  lang_code: lang_name
  for lang_name, lang_code in LANG_NAME_TO_CODE_GOOGLE.items()
}
```

Comment (GokulNC) [reply backend/users/utils.py:37]: Similarly for `LANG_NAME_TO_CODE_ULCA`

File: .gitignore (hunk 1 above)
Comment (GokulNC) [reply .gitignore:4]: If we really to need add a JSON or CSV file, we can force add by `git add -f <filename>`

File: backend/utils/lang_code_to_lang_name.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (GokulNC) [inline backend/utils/lang_code_to_lang_name.json:1]: Sorry, but where is this file being used? I don't see any refs to this.

Comment (GokulNC) [reply backend/utils/lang_code_to_lang_name.json:1]: Same comment for `backend/utils/lang_name_to_lang_code.json` as well

File: backend/functions/views.py [hunk 4]
```diff
@@ -1,30 +1,145 @@
...
+            context,
+            quality_status,
+            metadata,
+        ) in input_sentences:
+
+            # Only perform the saving if quality status is clean and corrected text is not empty
+            if (quality_status == "Clean") and (input_sentence is not None):
+
+                # Get the translations
+                translated_sentence = get_translation_using_cdac_model(
```
Comment (GokulNC) [inline backend/functions/views.py:92]: One request:

I think translating line-by-line will be quite time-consuming.
Can you please check once with Aravinth if there's an CDAC IndicTrans endpoint that supports batch translation of multiple sentences?

If that's supported, can you please ask him the optimal batch_size and do a batch-inference instead of sentence-by-sentence?

**Review by GokulNC — CHANGES_REQUESTED** [review]
I briefly glanced through the changes, I will go-through this in detail tomorrow.
Good night!

File: .gitignore (hunk 1 above)
Comment (prakharrathi25) [reply .gitignore:4]: Okay, makes sense. I will remove this json file and an exception for all of them. I didn't add these csv files to the `.gitignore` but I can remove them too and add an exception for them.

File: backend/users/utils.py (hunk 2 above)
Comment (prakharrathi25) [reply backend/users/utils.py:37]: Okay, I will do this. This is a helpful trick.

File: backend/functions/views.py (hunk 4 above)
Comment (prakharrathi25) [reply backend/functions/views.py:92]: We can discuss with @aravinth here directly about the batch translations in IndicTrans models.

File: backend/utils/lang_code_to_lang_name.json (hunk 3 above)
Comment (prakharrathi25) [reply backend/utils/lang_code_to_lang_name.json:1]: Ah, this was an earlier method that I was using but I went with the utils method then. I will remove these.

File: backend/functions/views.py (hunk 4 above)
Comment (aravinth) [reply backend/functions/views.py:92]: Hi @prakharrathi25 @GokulNC , 
Batching is supported in NMT. (input object is an array)
Ideal batch size = 50 to 75

image

Comment (prakharrathi25) [reply backend/functions/views.py:92]: Thanks, I will look at this.

Comment (prakharrathi25) [reply backend/functions/views.py]: Hi, @GokulNC, I have added batch translations to the `functions`. Please let me know if more changes are required.

File: backend/functions/utils.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    try:
+        response = requests.post(
+            "https://nmt-models.ulcacontrib.org/aai4b-nmt-inference/v0/translate",
+            headers=headers,
+            json=json_data,
+        )
+
+        output = ast.literal_eval(response.content.decode("utf-8"))
```
Comment (GokulNC) [inline backend/functions/utils.py:53]: How is this working without `import ast` ?

Also, `output = response.json()` should be sufficient right?

File: backend/functions/utils.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    # Create the input sentences list
+    input_sentences = [{"source": sentence} for sentence in sentence_list]
+
+    headers = {
+        # Already added when you pass json= but not when you pass data=
+        # 'Content-Type': 'application/json',
+    }
+
+    json_data = {
+        "input": input_sentences,
```
Comment (GokulNC) [inline backend/functions/utils.py:36]: Hey, Aravinth mentioned that we can only send a max of 75 sentences per requests. So we'll have to loop through the `input_sentences` and call the API.

Something like this:
```py
output_sentences = []
for i in range(0, len(input_sentences), batch_size):
  batch_of_input_sentences = input_sentences[i:i+batch_size]
  # call the API for `batch_of_input_sentences`
  # append the output translations to an array
```

File: backend/functions/views.py [hunk 7]
```diff
@@ -1,30 +1,141 @@
...
+            input_sentence = row["corrected_text"]
+            input_language = row["input_language"]
+            translated_sentence = row["translated_text"]
+            metadata = row["metadata"]
+            context = row["context"]
+
+            # Get the output dataset instance
+            output_dataset_instance = dataset_models.DatasetInstance.objects.get(
+                instance_id=output_dataset_instance_id
+            )
```
Comment (GokulNC) [inline backend/functions/views.py:95]: We already have this block outside the loop. Please remove.

**Review by GokulNC — CHANGES_REQUESTED** [review]

File: backend/functions/utils.py (hunk 5 above)
Comment (prakharrathi25) [reply backend/functions/utils.py:53]: Changed this to `response.json()`

File: backend/functions/utils.py (hunk 6 above)
Comment (prakharrathi25) [reply backend/functions/utils.py:36]: Right and I will set the batch size as 75 at the moment then.

File: backend/functions/views.py (hunk 7 above)
Comment (prakharrathi25) [reply backend/functions/views.py:95]: Done.

File: backend/functions/utils.py (hunk 6 above)
Comment (prakharrathi25) [reply backend/functions/utils.py:36]: Added this change as well. Please review.

File: backend/functions/utils.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        sentence_list (str): List of sentences to be translated.
+        source_language (str): Original language of the sentence.
+        target_language (str): Final language of the sentence.
+
+    Returns:
+        list: List of dictionaries containing the translated sentences.
+    """
+
+    # Get the translation model ID
+    model_id = LANG_TRANS_MODEL_CODES.get(f"{source_language}-{target_language}", 144)
```
Comment (GokulNC) [inline backend/functions/utils.py:79]: Instead of hardcoding `144`, can you please create a global variable, something like `DEFAULT_ULCA_INDIC_MODEL_ID` and use that in both the functions: `get_batch_translations_using_indictrans_nmt_api()` and `get_translation_using_cdac_model()`

File: backend/functions/utils.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    source_language = LANG_NAME_TO_CODE_ULCA[source_language]
+    target_language = LANG_NAME_TO_CODE_ULCA[target_language]
+
+    # Create the input sentences list
+    input_sentences = [{"source": sentence} for sentence in sentence_list]
+
+    headers = {
+        # Already added when you pass json= but not when you pass data=
+        # 'Content-Type': 'application/json',
+    }
```
Comment (GokulNC) [inline backend/functions/utils.py:91]: If this is not being used, can you please remove it.

File: backend/functions/views.py [hunk 10]
```diff
@@ -1,30 +1,178 @@
...
+        input_sentences_df = input_sentences_df[
+            (input_sentences_df["quality_status"] == "Clean")
+        ].reset_index(drop=True)
+
+        # Check if the dataframe is empty
+        if input_sentences_df.shape[0] == 0:
+            return {"error": "No clean sentences found. Perform project export first."}
+
+        # Make a sentence list for valid sentences to be translated
+        all_sentences_to_be_translated = input_sentences_df["corrected_text"].tolist()
```
Comment (GokulNC) [inline backend/functions/views.py:64]: Since this block of code is common for all langs, shall we move this outside the loop to avoid running this multiple times?

File: backend/functions/views.py [hunk 11]
```diff
@@ -1,30 +1,178 @@
...
+
+                # Check if translation output didn't return an error
+                if isinstance(translations_output, Exception):
+                    return translations_output
+
+                # Collect the translated sentences
+                translated_sentences = [
+                    translation["target"] for translation in translations_output
+                ]
+
```
Comment (GokulNC) [inline backend/functions/views.py:91]: Since we might add other services as well, to keep this function generic, shall we just move this inside `get_batch_translations_using_indictrans_nmt_api()` itself, and ensure that it returns an array of direct translations?

File: backend/functions/views.py [hunk 12]
```diff
@@ -1,30 +1,178 @@
...
+                )
+
+                # Check if translation output didn't return an error
+                if isinstance(translations_output, Exception):
+                    return translations_output
+
+                # Collect the translated sentences
+                translated_sentences = [
+                    translation["translatedText"] for translation in translations_output
+                ]
```
Comment (GokulNC) [inline backend/functions/views.py:107]: Same comment as above

File: backend/functions/views.py [hunk 13]
```diff
@@ -1,30 +1,178 @@
...
+            # Check if the translated sentences are equal to the input sentences
+            if len(translated_sentences) != len(batch_of_input_sentences):
+                return {
+                    "error": "The number of translated sentences does not match the number of input sentences."
+                }
+
+            # Get the output dataset instance
+            output_dataset_instance = dataset_models.DatasetInstance.objects.get(
+                instance_id=output_dataset_instance_id
+            )
```
Comment (GokulNC) [inline backend/functions/views.py:121]: Shall we also move this outside the loop to avoid repeated DB queries?

File: backend/functions/views.py [hunk 14]
```diff
@@ -1,30 +1,178 @@
...
+                    parent_data=sentence_text_object,
+                    instance_id=output_dataset_instance,
+                    input_language=input_language,
+                    output_language=output_language,
+                    input_text=input_sentence,
+                    machine_translation=translated_sentence,
+                    context=context,
+                    metadata_json=metadata,
+                )
+                translation_pair_obj.save()
```
Comment (GokulNC) [inline backend/functions/views.py:150]: Bulk create would be more cheaper and efficient I suppose. Can you append these to an array, and after all sentences in a batch is complete, can you do `bulk_create`

File: backend/functions/views.py [hunk 15]
```diff
@@ -111,37 +261,185 @@ def copy_from_ocr_document_to_block_text(request):
...
+    )
+
+    # Call the function to save the TranslationPair dataset
+    save_status = save_translation_pairs(
+        languages,
+        input_sentences,
+        output_dataset_instance_id,
+        batch_size=128,
+        api_type="google",
+    )
```
Comment (GokulNC) [inline backend/functions/views.py:357]: I suppose this would be a very time-consuming API call even for a dataset instance with 10k sentences. Shall we make this an async API?

Sorry I should have noticed this earlier. But I had mentioned in [SN-245](https://project-sunbird.atlassian.net/browse/SN-245) description that it should be a Celery task.

Can you please implement it that way?

File: backend/functions/views.py [hunk 16]
```diff
@@ -111,37 +261,185 @@ def copy_from_ocr_document_to_block_text(request):
...
+            instance_id=input_dataset_instance_id
+        ).values_list(
+            "id",
+            "corrected_text",
+            "language",
+            "context",
+            "quality_status",
+            "metadata_json",
+        )
+    )
```
Comment (GokulNC) [inline backend/functions/views.py:348]: If we are to make this a Celery task, we'll have to move this block into the `save_translation_pairs()` function.

Also, please rename the `save_translation_pairs` appropriately. (We are not just saving translations.)

**Review by GokulNC — CHANGES_REQUESTED** [review]
Please let me know if anything. Thanks!

File: backend/functions/utils.py (hunk 8 above)
Comment (prakharrathi25) [reply backend/functions/utils.py:79]: Sure, I will make that change.

File: backend/functions/views.py (hunk 11 above)
Comment (prakharrathi25) [reply backend/functions/views.py:91]: Do you want to do this for the Google Translate API as well? I wanted to keep the processing inside the api functions as minimum as possible but I can do it for both the functions then.

File: backend/functions/views.py (hunk 12 above)
Comment (prakharrathi25) [reply backend/functions/views.py:107]: Alright, I will make that change for both the functions in that case.

File: backend/functions/views.py (hunk 15 above)
Comment (prakharrathi25) [reply backend/functions/views.py:357]: Yeah, I haven't done that currently because that might involve a lot of breaking changes outside the `functions` module. Can we make a separate ticket for converting this to a celery task after this is merged to dev?

File: backend/functions/views.py (hunk 16 above)
Comment (prakharrathi25) [reply backend/functions/views.py:348]: I have renamed the function to `sentence_text_translate_and_save_translation_pairs()` and moved this block inside the function.

File: backend/functions/views.py (hunk 14 above)
Comment (prakharrathi25) [reply backend/functions/views.py:150]: Turns out `bulk_create` only works for a non-inherited model. I get the following error when I use it to create a TranslationPair model 

```
ValueError at /functions/schedule_google_translate_job
Can't bulk create a multi-table inherited model
```

I found some fixes on StackOverflow but they are mostly [user developed hacks](https://stackoverflow.com/questions/49826482/django-valueerror-cant-bulk-create-a-multi-table-inherited-model) and currently only work for single inheritance level, apparently. It might also be risky to implement them because they could lead to errors in the future.

File: backend/functions/views.py (hunk 11 above)
Comment (GokulNC) [reply backend/functions/views.py]: Yes please, it's better to keep API specific logic inside API specific utilities, and keep this function generic.

File: backend/functions/views.py (hunk 14 above)
Comment (GokulNC) [reply backend/functions/views.py:158]: > I found some fixes on StackOverflow

Thanks for sharing. Can we use this one?
https://stackoverflow.com/a/72939340/

Basically, shall we include this as a util, import it and use it wherever we want to `bulk_create` inherited models?
Do you see any drawbacks? (other than those mentioned below.

>  currently only work for single inheritance level
Yeah, even our models are mostly single inheritance level only.

> they could lead to errors in the future.
Yes, makes sense. But usually Django develops things in a way such that things are always backward-compatible.
So I feel it's fine. If required, if we face issues in next major version of Django, we can handle it.
What do you think?

File: backend/functions/views.py (hunk 15 above)
Comment (GokulNC) [reply backend/functions/views.py]: OK sure, makes sense. Can you maybe try the custom `bulk_create` thing (mentioned above) ?
Once that's done, you can request Ishvinder to test it out and merge.

You can also create a ticket yourself on Jira and assign to yourself for the Celery changes.

File: backend/functions/utils.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+    )
+
+    # Convert language names to the language code
+    source_language = LANG_NAME_TO_CODE_ULCA[source_language]
+    target_language = LANG_NAME_TO_CODE_ULCA[target_language]
+
+    headers = {
+        # Already added when you pass json= but not when you pass data=
+        # 'Content-Type': 'application/json',
+    }
```
Comment (GokulNC) [inline backend/functions/utils.py:141]: Please remove this as well. Missed it in the earlier review sorry

File: backend/functions/views.py (hunk 15 above)
Comment (GokulNC) [reply backend/functions/views.py]: Please also check if auto-increment works (the prev answer to that answer has mentioned it does not work)

File: backend/functions/views.py (hunk 14 above)
Comment (GokulNC) [reply backend/functions/views.py:158]: Please also check if auto-increment works (the prev answer to that answer has mentioned it does not work)

File: backend/functions/utils.py (hunk 17 above)
Comment (prakharrathi25) [reply backend/functions/utils.py:141]: Yeah, my bad. Missed this one previously.

File: backend/functions/views.py (hunk 15 above)
Comment (prakharrathi25) [reply backend/functions/views.py]: Sure, sounds good. I will try out the bulk create thing.

File: backend/functions/views.py (hunk 14 above)
Comment (prakharrathi25) [reply backend/functions/views.py]: I have tried this code. It's working and I have added the bulk create functionality. If we continue with single inheritance then I don't see any immediate issues. The autoincrement is also working since the TranslationPairs have been allotted appropriate IDs.

Comment (prakharrathi25) [reply backend/functions/views.py]: If this fine then I will ask @ishvindersethi22 to test and merge the code.

Comment (GokulNC) [reply backend/functions/views.py]: Thanks, looks good to me!

**Review by GokulNC — APPROVED** [review]

**Review by ishvindersethi22 — APPROVED** [review]
Tested on local. To be tested on prod after integration with env.

File: backend/functions/utils.py [hunk 18]
```diff
@@ -185,17 +193,25 @@ def get_translation_using_cdac_model(input_sentence, source_language, target_lan
...
+        checks_for_particular_languages (bool): If True, checks for the particular languages in the translations.
 
     Returns:
         list: List of dictionaries containing the translated sentences.
     """
 
+    if checks_for_particular_languages: 
+        # Checks for particular languages
+        if target_language in ["Bodo", "Maithili"]: 
+            target_language = "Hindi"
```
Comment (GokulNC) [inline backend/functions/utils.py:216]: I think Maithili is already supported in Google.
Can you please check once.

Comment (prakharrathi25) [reply backend/functions/utils.py:216]: @ishvindersethi22 requested this change. Can you please add your thoughts?

Comment (ishvindersethi22) [reply backend/functions/utils.py:216]: But the annotators need it in Hindi as per requirements.
